staging/src/k8s.io/cli-runtime/pkg/resource/query_param_verifier_v3.go

```diff
@@ -0,0 +1,153 @@
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+var _ Verifier = &queryParamVerifierV3{}
+
+// NewQueryParamVerifierV3 returns a pointer to the created queryParamVerifier struct,
+// which implements the Verifier interface.
+func NewQueryParamVerifierV3(dynamicClient dynamic.Interface, openAPIGetter discovery.OpenAPIV3SchemaInterface, queryParam VerifiableQueryParam) Verifier {
+	return &queryParamVerifierV3{
+		finder:        NewCRDFinder(CRDFromDynamic(dynamicClient)),
+		openAPIGetter: openAPIGetter,
+		queryParam:    queryParam,
+	}
+}
+
+// queryParamVerifierV3 encapsulates info necessary to determine if
+// the queryParam is a parameter for the Patch endpoint for a
+// passed GVK.
+type queryParamVerifierV3 struct {
+	finder        CRDFinder
+	openAPIGetter discovery.OpenAPIV3SchemaInterface
+	queryParam    VerifiableQueryParam
+}
+
+// HasSupport returns nil error if the passed GVK supports the parameter
+// (stored in struct; usually "fieldValidation") for Patch endpoint.
+// Returns an error if the passed GVK does not support the query param,
+// or if another error occurred.
+func (v *queryParamVerifierV3) HasSupport(gvk schema.GroupVersionKind) error {
+	// OpenAPI V3 client.
+	oapiv3Client := v.openAPIGetter.OpenAPIV3()
+	if oapiv3Client == nil {
+		return fmt.Errorf("failed to retrieve Open API V3 client")
+	}
+	// Downloads top-level Open API V3 document which contains a
+	// map of all API Server paths to Open API V3 GV URLs, including
+	// the hash for cache busting as query param.
+	//   Example: "apis/apps/v1: /openapi/v3/apis/apps/v1?hash=<HASH>"
+	paths, err := oapiv3Client.Paths()
+	if err != nil {
+		return fmt.Errorf("failed to fetch list of groupVersions: %w", err)
+	}
+	// From GVK, generate the key API Server path.
+	var resourcePath string
+	if len(gvk.Group) == 0 {
+		resourcePath = fmt.Sprintf("api/%s", gvk.Version)
+	} else {
+		resourcePath = fmt.Sprintf("apis/%s/%s", gvk.Group, gvk.Version)
+	}
+	// Returns the Group/Version Open API V3 URL from the paths map.
+	gv, exists := paths[resourcePath]
+	if !exists {
+		return fmt.Errorf("could not locate schema for %s", resourcePath)
+	}
+	// Download the Group/Version Open API V3 file (by content type).
+	openAPISchemaBytes, err := gv.Schema(runtime.ContentTypeJSON)
+	if err != nil {
+		return fmt.Errorf("failed to fetch openapi schema for %s: %w", resourcePath, err)
+	}
+	// Unmarshal the downloaded Group/Version bytes into the spec3.OpenAPI struct.
+	var parsedV3Schema spec3.OpenAPI
+	if err := json.Unmarshal(openAPISchemaBytes, &parsedV3Schema); err != nil {
+		return fmt.Errorf("failed to parse openapi schema for %s: %w", resourcePath, err)
+	}
```

Comment:
This is a lot of work to get the OpenAPI for a specific group version. And this will force re-deserializing every gv-openapi for every single object that we apply, which is not acceptable performance-wise.